src/Microsoft.IdentityModel.Tokens/Json/JsonSerializerPrimitives.cs

```diff
@@ -976,93 +976,97 @@ public static void WriteObject(ref Utf8JsonWriter writer, string key, object obj
 
             Type objType = obj.GetType();
 
-            if (obj is string str)
-                writer.WriteString(key, str);
-            else if (obj is long l)
-                writer.WriteNumber(key, l);
-            else if (obj is int i)
-                writer.WriteNumber(key, i);
-            else if (obj is bool b)
-                writer.WriteBoolean(key, b);
-            else if (obj is DateTime dt)
-                writer.WriteString(key, dt.ToUniversalTime().ToString("O", CultureInfo.InvariantCulture));
-            else if (typeof(IDictionary).IsAssignableFrom(objType))
+            switch (obj)
             {
-                IDictionary dictionary = (IDictionary)obj;
-                writer.WritePropertyName(key);
-
-                writer.WriteStartObject();
-                foreach (var k in dictionary.Keys)
-                    WriteObject(ref writer, k.ToString(), dictionary[k]);
-
-                writer.WriteEndObject();
-            }
-            else if (typeof(IList).IsAssignableFrom(objType))
-            {
-                IList list = (IList)obj;
-                writer.WriteStartArray(key);
-                foreach (var k in list)
-                    WriteObjectValue(ref writer, k);
-
-                writer.WriteEndArray();
-            }
-            else if (obj is JsonElement j)
-            {
-                writer.WritePropertyName(key);
-                j.WriteTo(writer);
-            }
-            else if (obj is double dub)
-                // Below net6.0, we have to convert the double to a decimal otherwise values like 1.11 will be serailized as 1.1100000000000001
-                // large and small values such as double.MaxValue and double.MinValue cannot be converted to decimal.
-                // In these cases, we will write the double as is.
+                case string str:
+                    writer.WriteString(key, str);
+                    break;
+                case long l:
+                    writer.WriteNumber(key, l);
+                    break;
+                case int i:
+                    writer.WriteNumber(key, i);
+                    break;
+                case bool b:
+                    writer.WriteBoolean(key, b);
+                    break;
+                case DateTime dt:
+                    writer.WriteString(key, dt.ToUniversalTime().ToString("O", CultureInfo.InvariantCulture));
+                    break;
+                case IDictionary dictionary:
+                    writer.WritePropertyName(key);
+                    writer.WriteStartObject();
+                    foreach (var k in dictionary.Keys)
+                        WriteObject(ref writer, k.ToString(), dictionary[k]);
+                    writer.WriteEndObject();
+                    break;
+                case IEnumerable enumerable:
```

Comment:
All the logic is the same in this switch statement except here I switched it from IList to IEnumerable which should be fine since the only thing it is doing is running through it with a foreach

Reply:
IEnumerable is slower than IList.
Why the switch?

Reply:
Partly because [the guidance here](https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/guidelines-for-collections#collection-parameters) suggests when choosing between collections, choose the least specific option that can do what you need which in this case all we need is to enumerate.

In the SecurityTokenDescriptor after discussing with JM and Keegan we went with IEnumerable for the public Audiences member to match other public fields we have as well as be in line with the above guidance. If I didn't change this I'd have to be calling ToList() on the Audiences collection every time we want to serialize it.

Reply:
When I tried changing it to IList it broke something in the .NET Core / Wilson GitHub Action Test (pull_request) that I can't repro locally... so leaving it for now.

Reply:
I disagree with the IEnumerable change.
I think it is good guidance for some cases, but when we are creating billions of tokens a day, performance and allocations are more important.

Reply:
changed it back